ddtrace/contrib/openai/_endpoint_hooks.py

```diff
@@ -3,25 +3,68 @@
 
 from ..trace_utils import set_flattened_tags
 from .utils import _est_tokens
+from .utils import _format_bool
+from .utils import _format_openai_api_key
 
 
-class _EndpointHook:
-    _request_tag_attrs = []
-    _default_name = "openai"
+API_VERSION = "v1"
+
 
-    def _record_request(self, span, kwargs):
-        for kw_attr in self._request_tag_attrs:
+class _EndpointHook:
+    # Assume base arg signature follows openai.EngineAPIResource.create(...)
+    _request_arg_params = ("cls", "api_key", "api_base", "api_type", "request_id", "api_version", "organization")
+    _request_kwarg_params = ()
+    _prompt_completion = False
+    ENDPOINT_NAME = "openai"
+    REQUEST_TYPE = ""
+    OPERATION_ID = ""
+
+    def _record_request(self, span, integration, args, kwargs):
+        """Set base-level openai tags, as well as request params from args and kwargs."""
+        endpoint = self.ENDPOINT_NAME
+        if endpoint is None:
+            endpoint = "%s" % args[0].OBJECT_NAME
+        span.set_tag_str("openai.request.endpoint", "/%s/%s" % (API_VERSION, endpoint))
+        span.set_tag_str("openai.request.method", self.REQUEST_TYPE)
+
+        base_level_tag_args = ("api_base", "api_type", "api_version")
+
+        if self._request_arg_params and len(self._request_arg_params) > 1:
```

Comment:
The problem with that approach is that there is so much inconsistency in the openai-python library in regards to what args/kwargs are common to different endpoints, as well as whether the same arg is a kwarg for different endpoints, and common methods for certain endpoints. Ideally I would go with your approach (and that was my first approach at this), but the inconsistency of openai-python would result in a *lot* of duplicate code.

I think for now we can stick with this, and if their library improves in consistency, then we can consider shifting the design like you mentioned.

Reply:
On second thought, I think this might increase maintainability if we follow your suggestion. I've changed the EndpointHook design such that all EndpointHooks need to call the super method, which will automatically tag all text/numeric/dict request params, and each endpoint hook will need to handle all special case params.